pageserver/src/task_mgr.rs

```diff
@@ -0,0 +1,457 @@
+//! Tokio select!().
+//!
+//!
+//! TODO: This would be a good place to also handle panics in a somewhat sane way.
+//! Depending on what task panics, we might want to kill the whole server, or
+//! only a single tenant or timeline.
+//!
+
+use std::collections::HashMap;
+use std::future::Future;
+use std::panic::AssertUnwindSafe;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+use futures::FutureExt;
+use tokio::runtime::Runtime;
+use tokio::sync::watch;
+use tokio::task::JoinHandle;
+use tokio::task_local;
+
+use tracing::{debug, error, info, warn};
+
+use once_cell::sync::Lazy;
+
+use utils::zid::{ZTenantId, ZTimelineId};
+
+use crate::shutdown_pageserver;
+
+//
+// There are four runtimes:
+//
+// Compute request runtime
+//  - used to handle connections from compute nodes. Any tasks related to satisfying
+//    GetPage requests, base backups, import, and other such compute node operations
+//    are handled by the Compute request runtime
+//  - page_service.rs
+//  - this includes layer downloads from remote storage, if a layer is needed to
+//    satisfy a GetPage request
+//
+// Management request runtime
+//  - used to handle HTTP API requests
+//
+// WAL receiver runtime:
+//  - used to handle WAL receiver connections.
+//  - and to receiver updates from etcd
+//
+// Background runtime
+//  - layer flushing
+//  - garbage collection
+//  - compaction
+//  - remote storage uploads
+//  - initial tenant loading
+//
+// Everything runs in a tokio task. If you spawn new tasks, spawn it using the correct
+// runtime.
+//
+// There might be situations when one task needs to wait for a task running in another
+// Runtime to finish. For example, if a background operation needs a layer from remote
+// storage, it will start to download it. If a background operation needs a remote layer,
+// and the download was already initiated by a GetPage request, the background task
+// will wait for the download - running in the Page server runtime - to finish.
+// Another
+// example: the initial tenant loading tasks are launched in the background ops runtime.
+// If a GetPage request comes in before the load of a tenant has finished, the GetPage
+// request will wait for the tenant load to finish.
+//
+// The core Timeline code is synchronous, and uses a bunch of std Mutexes and RWLocks to
+// protect data structures. Let's keep it that way. Synchronous code is easier to debug
+// and analyze, and there's a lot of hairy, low-level, performance critical code there.
+//
+// It's nice to have different runtimes, so that you can quickly eyeball how much CPU
+// time each class of operations is taking, with 'top -H' or similar.
+//
+// It's also good to avoid hogging all threads that would be needed to process
+// other operations, if the upload tasks e.g. get blocked on locks. It shouldn't
+// happen, but still.
+//
+pub static COMPUTE_REQUEST_RUNTIME: Lazy<Runtime> = Lazy::new(|| {
```

Comment:
I think we already can hide a few of those runtimes and eventually should, but that's a plan for future PRs.
The idea would be to have a few task-specific runtimes isolated and 100% managed by the task_mgr or similar, and the rest of the code just doing `.await` on `async fn`.

Reply:
I didn't quite understand what you have in mind. But since it's a plan for the future, that's OK for now :-).